app/api/sul_bib/authorship_api.rb

```diff
@@ -201,34 +158,38 @@ def get_publication(sul_pub_id, pmid, sw_id)
       requires :visibility, type: String, desc: 'The JSON body must indicate if the contribution is visible', coerce_with: ->(val) { val.downcase }
     end
     post do
-      request_body_unparsed = env['api.request.input']
       logger.info('POST Contribution JSON: ')
-      logger.info(request_body_unparsed.to_s)
+      logger.info(env['api.request.input'].to_s)
 
       # Find or create an author
       author = get_author(
         params[:cap_profile_id],
         params[:sul_author_id]
       )
-      check_author_ids(author, params[:cap_profile_id])
+      check_author_ids!(author, params[:cap_profile_id])
+
+      ids = params.slice(:sul_pub_id, :pmid, :sw_id, :wos_uid).to_h.symbolize_keys
+      ids.reject! { |_, v| v.blank? }
+      unless ids.any?
+        log_and_error!('There is no valid publication identifier: sul_pub_id || pmid || sw_id || wos_uid.', 400)
+      end
 
       # Now find an existing sul publication or, if it doesn't exist, it
-      # may be fetched from PubMed (pmid) or ScienceWire (sw_id).
-      pub = get_publication(
-        params[:sul_pub_id],
-        params[:pmid],
-        params[:sw_id]
-      )
+      # may be fetched from PubMed (pmid), WebOfScience (wos_uid) or ScienceWire (sw_id).
+      pub = get_local_publication!(ids[:sul_pub_id]) if ids[:sul_pub_id]
+      pub ||= get_publication_via_pubmed!(ids[:pmid]) if ids[:pmid]
+      pub ||= get_publication_via_sciencewire!(ids[:sw_id]) if ids[:sw_id]
+      pub ||= get_publication_via_wos!(author, ids[:wos_uid]) if ids[:wos_uid]
```

Comment:
Maybe each of the `get_publication_*` could use some exception handling that would allow this block to move on to the next option?  Even if it's not our preferred source, if we can get it - good.  It looks like one source throwing an exception could break the chain here.

Reply:
It's not exception handling, we intentionally respond w/ `404` on the first id that doesn't work out.  I think this behavior is correct, given that the API is expecting a **known item** to be referenced, something that was *already* retrieved previously.  Also given that performance will degrade if we are serially searching N remote backends instead one.

The possibility of receiving multiple IDs of different types is an artifact of each param being optional, not something that had intentional use.  So it was barely unexplored by the existing implementation (only via the priority order of the backends).  No rectification was performed (e.g., if we receive 2 ids, does the retrieved object match both of them?), and I wouldn't really want to start that now.

It probably would have been better to use a single required `uid` param w/ namespace.  The cost is that it would have required some param validation and interpretation, but this question would have been effectively excluded.  Also would have been easier to extend for new DBs, since we wouldn't have to add a new param.

Reply:
Hmmm, yea, not sure about scope here.  On the one hand, this API should be mostly used, if not exclusively, for associating or updating authorship for an existing pub.  BUT, that's not entirely true.  It _does_ allow harvests too.  It's the kitchen sink of "hey, show me the pub that I own" (given one or more pub ids).  I'm not entirely clear about the assumption that this API is only used with a known pub-id that was retrieved from us before it came back to us.

Reply:
I think this amounts to exactly where the `log_and_error` is called and whether it should wrap the entire block of options to find/harvest a pub or if it should be in each one of them.  As it is now, it's in each one of them and any one of them will break the chain.

Reply:
See https://github.com/sul-dlss/sul_pub/commit/9decc0fc96903f95b76ba38c49fa873750bebcd0

Reply:
You misinterpret me.  The item **known**, that doesn't mean it is local.  Title search is not accomplished via authorship API.  

Your commit obscures a range of possible different errors with the same message.